SubscribeMulti + UnsubscribeMulti, multiplicities

kazimuth opened the pull request:
## Description of Changes
Add SubscribeMulti and UnsubscribeMulti from upstream.
Fix unsubscribe bug found by testing against Bitcraft: Calling `UnsubscribeThen` with any (non-null) callback would result in an exception incorrectly telling the user that `Unsubscribe` had been called twice.
Multiplicity support is implemented with unit tests + manual testing of quickstart-chat.
Also, rows in the client cache are now looked up by primary key if available, which I suspect is going to be a large performance boost.

## API

 - [x] This is an API breaking change to the SDK

*If the API is breaking, please state below what will break*

## Requires SpacetimeDB PRs
*List any PRs here that are required for this SDK change to work*

## Testsuite
*If you would like to run the your SDK changes in this PR against a specific SpacetimeDB branch, specify that here. This can be a branch name or a link to a PR.*

SpacetimeDB branch name: jgilles/final-cs-codegen-changes

## Testing
*Write instructions for a test that you performed for this PR*

- [x] unit tests
- [ ] Working with Ryan to add a test suite.
- [x] blackholio

kazimuth committed:
Fix unsubscribe bug found by testing against Bitcraft

Warning message fix (thanks Jeff)

Update for new Subscribe/UnsubscribeMulti

thanks zeke

Add MultiDictionary; update dispatching logic still wrong

Add even more wacky data structures

WHOOOOOO

Fix MultiDictionary bug

More asserts and comments

Tests passing

Better errors, purge dead package

Test

Update for blackholio

Update blackholio

Fix comment

Remove dead

One more dead

COMMENT

More comments

Added a multiplicity test example

Creates a small sample program with a Rust server and C# client that tests various subscriptions and outputs the results to the CLI.

Make sure we insert before we delete

Fix more comments

Add CRDT test for MultiDictionaryDelta

Revert "Added a multiplicity test example"

This reverts commit ff109257e62f57542cecd0a64dda44178ea5d419.

Undo DLL changes

bfops commented:
I changed the base branch to `master`; I'll re-cut the release branch when things settle down.

bfops commented:
Out of curiosity, why did the DLLs need updating?

bfops commented on src/Event.cs:
```diff
@@ -254,7 +254,7 @@ public void UnsubscribeThen(Action<SubscriptionEventContext>? onEnded)
                 throw new Exception("Cannot unsubscribe from inactive subscription.");
             }
-            if (onEnded != null)
+            if (this.onEnded != null)
```
it also looks like we have a separate "bug" where calling `Unsubscribe` twice won't actually trigger the exception below? (Because `this.onEnded` will be `null` if we called `Unsubscribe` since that triggers this function with a `null` param) Is that.. an issue..?

kazimuth replied:
GOOOD point

kazimuth commented:
Blackholio tested on https://github.com/clockworklabs/Blackholio/pull/22

kazimuth commented:
@bfops uhh what branch should this be based on?

joshua-spacetime commented:
@kazimuth if I'm reading this correctly, inserts are applied to the cache before deletes, correct?

kazimuth commented:
Good catch -- but that loop is accumulating into a MultiDictionaryDelta, which is a collection of key-value pairs with signed multiplicities. Since the multiplicities can go negative, you're allowed to remove a row before you insert it; in general, the order Adds and Removes are applied to a MultiDictionaryDelta doesn't matter. The resulting delta will have the same effect when Applied to a MultiDictionary. The proptest ShuffleDelta tests this.

I should definitely comment about this, thank you.

kazimuth commented on src/MultiDictionary.cs:
```diff
@@ -0,0 +1,416 @@
+
+    }
+
+    /// <summary>
+    /// A bulk change to a multidictionary. Allows both adding and removing rows.
+    /// 
+    /// Can be applied to a multidictionary, and also inspected before application to see
+    /// what rows will be deleted. (This is used for OnBeforeDelete.)
+    /// 
+    /// Curiously, the order of operations applied to a MultiDictionaryDelta does not matter.
+    /// No matter the order of Add and Remove calls on a delta, when the Delta is applied,
+    /// the result will be the same, as long as the Add and Remove *counts* for each KeyValuePair are
+    /// the same.
+    /// (This means that this is a "conflict-free replicated data type", unlike MultiDictionary.)
```
@joshua-spacetime

joshua-spacetime replied:
Excellent, thanks @kazimuth

kazimuth commented:
@bfops , possibly the DLLs did NOT need updating and I did it unnecessarily. I just ran my update-everything script out of habit, but I can undo that.

Update: Done.

kazimuth commented on examples~/multiplicity/client/Program.cs:
```diff
@@ -0,0 +1,526 @@
+﻿// Multiplicity Test Client
+// This test adds several dogs to the Multiplicity Test Server,
+// stores a local model of those dogs,
+// creates several connections in different combinations,
```
Thanks for the test @rekhoff !

```suggestion
// creates several subscriptions in different combinations,
```

kazimuth commented on examples~/multiplicity/client/nuget.config:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Ah, this file should be .gitignored.

kazimuth commented:
@rekhoff I'm wondering if we should put that multiplicity test into its own PR. I think we probably want to merge this and add that test later. It will require some git-fu I can help out with.

kazimuth commented on examples~/multiplicity/client/Model.cs:
```diff
@@ -0,0 +1,164 @@
+    public HashSet<Cat> ExpectedServerCats = new HashSet<Cat>();
+    
+    public HashSet<Dog> ExpectedClientDogs = new HashSet<Dog>();
+    public HashSet<Cat> ExpectedClientCats = new HashSet<Cat>();
```
If you want these in the model, you'll want e.g.:

```csharp
List<Func<Model, HashSet<Dog>>> activeDogSubscriptions = new();
public HashSet<Dog> ExpectedClientDogs {
    get {
        HashSet<Dog> result = new();
        foreach (var activeDogSub in activeDogSubscriptions) {
           result.UnionWith(activeDogSub(this));
        } 
        return result;
     }
}
```

kazimuth commented:
@rekhoff, I moved your code to the branch `rekhoff/multiplicity-example` and rebased it. Would appreciate if you could open a PR with that code based at this branch.

kazimuth committed:
Simplify ProcessedDatabaseUpdates construction

kazimuth committed:
A few more comments

gefjon commented:
John: What's the story on the subscription helper?
James: That can go away now. Also, I want to test this against BitCraft.

rekhoff commented:
> @rekhoff, I moved your code to the branch `rekhoff/multiplicity-example` and rebased it. Would appreciate if you could open a PR with that code based at this branch.

[PR 254](https://github.com/clockworklabs/com.clockworklabs.spacetimedbsdk/pull/254) opened for multiplicity test.

kazimuth committed:
Update for SpacetimeDB jgilles/final-cs-codegen-changes

kazimuth committed:
Rerun tests

kazimuth commented:
Now based on https://github.com/clockworklabs/SpacetimeDB/pull/2303

kazimuth committed:
Better variable names and doc comments in MultiDictionary

kazimuth committed:
Update README

rekhoff commented on examples~/quickstart/client/Program.cs:
```diff
@@ -123,25 +124,15 @@ void OnConnect(DbConnection conn, Identity identity, string authToken)
-        }
-    };
-
     var userSubscription = conn.SubscriptionBuilder()
```
This now contains both a user and message subscription, to variable maybe wants to be changed from `userSubscription` to be more clear.

kazimuth committed:
Remove dead internal API

kazimuth commented on tests~/MultiDictionaryTests.cs:
```diff
@@ -0,0 +1,239 @@
+    {
+        ListWithNormalizedDuplicates(Gen.Byte[1, 10], Gen.Byte[1, 10], EqualityComparer<byte>.Default).Sample(list =>
+        {
+            var m1 = MultiDictionary<byte, byte>.FromEnumerable(list, EqualityComparer<byte>.Default, EqualityComparer<byte>.Default);
+            Gen.Shuffle(list);
+            var m2 = MultiDictionary<byte, byte>.FromEnumerable(list, EqualityComparer<byte>.Default, EqualityComparer<byte>.Default);
+
+            Assert.Equal(m1, m2);
```
The code inside `Sample` is what actually matters in these tests. Basically, `list` is a list of `KeyValuePair<byte, byte>` that is randomly generated, and we test that no matter what value it has, this code works.

kazimuth commented on tests~/MultiDictionaryTests.cs:
```diff
@@ -0,0 +1,239 @@
+    /// <param name="g2"></param>
+    /// <param name="equality"></param>
+    /// <returns></returns>
+    Gen<List<KeyValuePair<TKey, TValue>>> ListWithNormalizedDuplicates<TKey, TValue>(Gen<TKey> g1, Gen<TValue> g2, IEqualityComparer<TKey> equality, int ListMinLength = 0, int ListMaxLength = 32)
```
The only thing that really matters about this is that it returns `Gen<List<KeyValuePair<TKey, TValue>>>` which is an object that randomly generates `List<KeyValuePair<TKey, TValue>>`s.

kazimuth committed:
Better test comments

kazimuth committed:
Fix variable name

rekhoff reviewed (approved):
Reviewed code. Looks correct and well commented. No issues found.

joshua-spacetime commented on src/MultiDictionary.cs:
```diff
@@ -0,0 +1,434 @@
+                if (self.RawDict.TryGetValue(entry.Key, out var mine))
+                {
+                    var resultMultiplicity = (int)mine.Multiplicity + theirDelta;
+                    return resultMultiplicity <= 0;
```
Note, it should never be the case that `resultMultiplicity < 0`. I would at least log it if it does happen.

joshua-spacetime commented on src/MultiDictionary.cs:
```diff
@@ -0,0 +1,434 @@
+                        // This is a removal.
+                        if (newMultiplicity < 0)
+                        {
+                            PseudoThrow($"Internal error: Removing row with key {key} {-theirDelta} times, but it is only present {my.Multiplicity} times.");
```
Just like you're doing here actually 👍

joshua-spacetime reviewed (approved):
I've reviewed the MultiDictionary and Table cache changes for tracking multiplicities and invoking row callbacks. Just a small logging nit, but otherwise the logic is correct.

As for the rest of the changes, I'll defer to @rekhoff.
